File: dispatch/with_data/Julia/dispatch_switch.jl
# Define an enum to represent the type of data stored
@enum DataType TYPE_INT TYPE_FLOAT TYPE_DOUBLE TYPE_UNKNOWN

# Define a struct to hold the data and its type
mutable struct Value
    type::DataType
    data::Union{Int, Float32, Float64}
end

# Function to perform arithmetic on an int
process_int(value::Int) = value + 10

# Function to perform arithmetic on a float
process_float(value::Float32) = value * 2.0f0

# Function to perform arithmetic on a double
process_double(value::Float64) = value / 2.0

# Function to process the value based on its type
function process_value(value::Value)
    if value.type == TYPE_INT
        value.data = process_int(value.data)
    elseif value.type == TYPE_FLOAT
        value.data = process_float(value.data)
    elseif value.type == TYPE_DOUBLE
        value.data = process_double(value.data)
    else
        println("Unsupported type")
    end
    return value
end

# Function to read the data type from the config file
function read_config(filename::String)
    try
        type_str = readline(filename)
        if type_str == "int"
            return TYPE_INT
        elseif type_str == "float"
            return TYPE_FLOAT
        elseif type_str == "double"
            return TYPE_DOUBLE
        else
            return TYPE_UNKNOWN
        end
    catch e
        println("Failed to read config file: ", e)
        return TYPE_UNKNOWN
    end
end

function main()
    # Read the data type from the config file
    config_file = "config.txt"
    type = read_config(config_file)

    if type == TYPE_UNKNOWN
        println("Invalid or unsupported data type in config file")
        return
    end

    # Initialize the value based on the type
    value = if type == TYPE_INT
        Value(type, 42)
    elseif type == TYPE_FLOAT
        Value(type, 3.14f0)
    elseif type == TYPE_DOUBLE
        Value(type, 2.718281828459045)
    else
        error("Unsupported type")
    end

    process_value(value)

    # Start the clock
    start_time = time()

    # Process the value 1,000,000,000 times
    iterations = 1_000_000_000
    for _ in 1:iterations
        value = process_value(value)
    end

    # Stop the clock
    end_time = time()

    # Calculate the elapsed time in seconds
    elapsed_time = end_time - start_time

    # Print the result and the time taken
    println("Processed ", value.type, ": ", value.data)
    println("Time taken for ", iterations, " iterations: ", elapsed_time, " seconds")
end

# Run the program
main()

File: dispatch/with_data/C/dispatch_switch.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

// Define an enum to represent the type of data stored in the union
typedef enum {
    TYPE_INT,
    TYPE_FLOAT,
    TYPE_DOUBLE,
    TYPE_UNKNOWN
} DataType;

// Define a union to hold different types of data
typedef union {
    int intValue;
    float floatValue;
    double doubleValue;
} Data;

// Define a struct that holds the union and the type of data it contains
typedef struct {
    DataType type;
    Data data;
} Value;

// Function to perform arithmetic on an int
int processInt(int value) {
    return value + 10; // Example: add 10 to the int
}

// Function to perform arithmetic on a float
float processFloat(float value) {
    return value * 2.0f; // Example: multiply the float by 2
}

// Function to perform arithmetic on a double
double processDouble(double value) {
    return value / 2.0; // Example: divide the double by 2
}

// Function to process the value based on its type
Data processValue(Value *value) {
    switch (value->type) {
        case TYPE_INT:
            value->data.intValue = processInt(value->data.intValue);
            break;
        case TYPE_FLOAT:
            value->data.floatValue = processFloat(value->data.floatValue);
            break;
        case TYPE_DOUBLE:
            value->data.doubleValue = processDouble(value->data.doubleValue);
            break;
        default:
            printf("Unsupported type\n");
            break;
    }
}

// Function to read the data type from the config file
DataType readConfig(const char* filename) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        perror("Failed to open config file");
        return TYPE_UNKNOWN;
    }

    char type[16];
    if (fscanf(file, "%15s", type) != 1) {
        fclose(file);
        return TYPE_UNKNOWN;
    }

    fclose(file);

    if (strcmp(type, "int") == 0) {
        return TYPE_INT;
    } else if (strcmp(type, "float") == 0) {
        return TYPE_FLOAT;
    } else if (strcmp(type, "double") == 0) {
        return TYPE_DOUBLE;
    } else {
        return TYPE_UNKNOWN;
    }
}

int main() {
    // Read the data type from the config file
    const char* configFile = "config.txt";
    DataType type = readConfig(configFile);

    if (type == TYPE_UNKNOWN) {
        fprintf(stderr, "Invalid or unsupported data type in config file\n");
        return 1;
    }

    // Initialize the value based on the type
    Value value;
    value.type = type;

    switch (type) {
        case TYPE_INT:
            value.data.intValue = 42; // Example int value
            break;
        case TYPE_FLOAT:
            value.data.floatValue = 3.14f; // Example float value
            break;
        case TYPE_DOUBLE:
            value.data.doubleValue = 2.718281828459045; // Example double value
            break;
        default:
            break;
    }

    // Start the clock
    clock_t start = clock();

    // Process the value 1,000,000,000 times
    const int iterations = 1000000000;
    for (int i = 0; i < iterations; i++) {
        processValue(&value);
    }

    // Stop the clock
    clock_t end = clock();

    // Calculate the elapsed time in seconds
    double elapsedTime = (double)(end - start) / CLOCKS_PER_SEC;

    // Print the result and the time taken
    switch (type) {
        case TYPE_INT:
            printf("Processed int: %d\n", value.data.intValue);
            break;
        case TYPE_FLOAT:
            printf("Processed float: %f\n", value.data.floatValue);
            break;
        case TYPE_DOUBLE:
            printf("Processed double: %lf\n", value.data.doubleValue);
            break;
        default:
            break;
    }
    printf("Time taken for %d iterations: %.6f seconds\n", iterations, elapsedTime);

    return 0;
}
